airbyte-integrations/bases/base-python/base_python/cdk/utils/event_timing.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+
+import datetime
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from threading import Lock, local
+
+from base_python.logger import AirbyteLogger
+
+logger = AirbyteLogger()
+
+
+class EventTimer:
+    """Simple nanosecond resolution event timer for debugging, initially intended to be used to record streams execution
+    time for a source.
+       As it is likely that streams could be read in parallel, we use a thread local storage to store the
+       stack of events and thus keep track of the current event.
+       Event nesting follows a LIFO pattern, so finish will apply to the last started event.
+    """
+
+    thread_local = local()
+
+    def __init__(self, name):
+        self.name = name
+        self.events = {}
+        self.count = 0
+        EventTimer.thread_local.stack = []
+        self.stack = EventTimer.thread_local.stack
+        self.lock = Lock()
+
+    def start_event(self, name):
+        """
+        Start a new event and push it to the stack.
+        """
+        self.lock.acquire()
```

Comment:
why you need thread synchronization? airbyte cdk is not using and planning to use multi threading.

Reply:
Removed it